http/server_sent_event.ts

```diff
@@ -0,0 +1,411 @@
+ *   }, 2000);
+ *
+ *   target.addEventListener("close", () => clearInterval(id));
+ *   return target.asResponse();
+ * }, { port: 8000 });
+ * ```
+ *
+ * @module
+ */
+
+import { assert } from "../_util/asserts.ts";
+
+const encoder = new TextEncoder();
+
+const DEFAULT_KEEP_ALIVE_INTERVAL = 30_000;
+
+export interface ServerSentEventInit extends EventInit {
+  /** An optional `id` which will be sent with the event and exposed in the
+   * client `EventSource`. */
+  id?: number;
+
+  /** The replacer is passed to `JSON.stringify` when converting the `data`
+   * property to a JSON string. */
+  replacer?:
+    | (string | number)[]
+    // deno-lint-ignore no-explicit-any
+    | ((this: any, key: string, value: any) => any);
+
+  /** Space is passed to `JSON.stringify` when converting the `data` property
+   * to a JSON string. */
+  space?: string | number;
+}
+
+export interface ServerSentEventTargetOptions {
+  /** Keep client connections alive by sending a comment event to the client
+   * at a specified interval.  If `true`, then it polls every 30000 milliseconds
+   * (30 seconds). If set to a number, then it polls that number of
+   * milliseconds.  The feature is disabled if set to `false`.  It defaults to
+   * `false`. */
+  keepAlive?: boolean | number;
+}
+
+class CloseEvent extends Event {
+  constructor(eventInit: EventInit) {
+    super("close", eventInit);
+  }
+}
+
+/** An event which contains information which will be sent to the remote
+ * connection and be made available in an `EventSource` as an event. A server
+ * creates new events and dispatches them on the target which will then be
+ * sent to a client.
+ *
+ * See more about Server-sent events on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events)
+ *
+ * ### Example
+ *
+ * ```ts
+ * import {
+ *   ServerSentEvent,
+ *   SSEStreamTarget,
+ * } from "https://deno.land/std@$STD_VERSION/http/server_sent_event.ts";
+ * import { serve } from "https://deno.land/std@$STD_VERSION/http/server.ts";
+ *
+ * await serve((request) => {
+ *   const target = new SSEStreamTarget();
+ *   const evt = new ServerSentEvent(
+ *     "message",
+ *     { hello: "world" },
+ *     { id: 1 },
+ *   );
+ *   target.dispatchEvent(evt);
+ *   return target.asResponse();
+ * }, { port: 8000 });
+ * ```
+ */
+export class ServerSentEvent extends Event {
+  #data: string;
+  #id?: number;
+  #type: string;
+
+  /**
+   * @param type the event type that will be available on the client. The type
+   *             of `"message"` will be handled specifically as a message
+   *             server-side event.
+   * @param data  arbitrary data to send to the client, data this is a string
+   *              will be sent unmodified, otherwise `JSON.parse()` will be used
+   *              to serialize the value
+   * @param eventInit initialization options for the event
+   */
+  constructor(
+    type: string,
+    data: unknown,
```

Comment:
How about including `data` in `ServerSentEventInit`?

`CustomEvent` has similar structure as this class and it has `details` property in the init option bag. https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent

Reply:
Good point, also on the `MessageEvent`, which is the client side event type, it's constructor takes `data` in the options bag. I will rework it.